Use FQDN regex LRU everywhere

christarazi opened the pull request:
- defaults: Define FQDN regex compile LRU size default
- fqdn/re: Provide new package for accessing compile regex objects
- fqdn: Use new regex LRU package everywhere
- daemon, fqdn: Provide method for non-cached validation

See commit msgs.

Related: https://github.com/cilium/cilium/pull/17224
Related: https://github.com/cilium/cilium/pull/19383

christarazi committed:
fqdn/re: Provide new package for accessing compiled regex objects

This new package will be used throughout Cilium to fetch compiled regex
objects when handling toFQDN and L7 DNS rule names or patterns. The goal
of this package is to provide a seamless, transparent interface to
compile regexes or fetch them from the LRU.

Upcoming commits will utilize this new package for all potential regex
compilations in order to reduce CPU and memory consumption when
compiling regexes constantly.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
fqdn: Use new regex LRU package everywhere

This reduces the potential CPU and memory spikes that were observed with
constantly compiling toFQDN and L7 DNS rules, namely MatchName and
MatchPattern.

The spikes were observed in a high churn environment where many
FQDN-related policies were being added and deleted, many of the added
rules had the same MatchName and MatchPattern fields.

Manual testing was shown (by taking a pprof) that with 1024 regexes
present in the cache, it only consumed 10s of MB.

This commit ensures that we are always caching the regex compiled
objects, both for policy validation and when processing the policy /
selectors.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/fqdn/matchpattern/matchpattern.go:
```diff
@@ -24,7 +25,7 @@ func Validate(pattern string) (matcher *regexp.Regexp, err error) {
 	}
 
-	return regexp.Compile(ToRegexp(pattern))
+	return re.CompileRegex(ToRegexp(pattern))
```
Maybe worth calling it out here, but reviewers might wonder: how do we know if this will even use the LRU, i.e. if the LRU even initialized at the time this code path executes?

The answer is yes it will use the LRU considering that this code path is called to "sanitize" the policies (read validate) upon submission from the user.

The LRU is initialized during dameon bootstrap of the FQDN subsystem which happens before the K8s watcher sync starts (`InitK8sSubsystem()`), so before Cilium can even process the policies applied by the user.

christarazi replied:
Happy to add this as a comment somewhere or maybe inside the commit msg, once CI finishes.

joestringer replied:
Ah hehe, exactly the topic of my feedback below: https://github.com/cilium/cilium/pull/19632#discussion_r862192900

joestringer commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,55 @@
+// doesn't exist in the cache, it will be added.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	if RegexCompileLRU == nil {
+		return regexp.Compile(p)
```
Did you consider returning an error here?

I suspect that we never intend for this path to really be used, it's more of a programming error that indicates `InitRegexCompileLRU()` was not called early enough. Returning an error could help to identify this case and ensure that we just move the initialization earlier.

christarazi replied:
This is the area that I kinda struggled to come to a conclusion on. What I opted for here at a high-level is to ignore that initialization didn't happen and simply pass-through straight into a regex compile, transparent to the caller. And that was the biggest motivation--to be transparent. But with that comes that cost of hiding when a human error was made.

I just realized while typing this out that we could have a sync.Once inside this case to initialize the cache if it wasn't, so that should catch any "premature" calls to `CompileRegex` that were done without initializing the LRU first (in case of human error or even otherwise).

I'll implement that, thanks.

christarazi replied:
Hm, the initialization function needs to take an int for the size which we don't necessarily have with `CompileRegex` and would be awkward to always pass the size of the LRU to it... :thinking:

christarazi replied:
Fixed by using option.Config.FQDNRegexCompileLRUSize. If the init fails, we'll log a warning. I don't expect to hit this case going forward as the initialization happens during the DNS proxy bootstrap. Assuming the shuffling of subsystem inits don't occur, this should never happen. If it does happen, we'll have a warning log. LMK if you think this addresses the concern or if we should think about a deeper approach to the problem.

joestringer replied:
Minor nit here, this now initializes the LRU on the first pass, but if the initialization is successful it will actually not cache that one first compiled regex. That said, this is also not such a big deal since if the same regex gets used many times, then it will end up in the LRU soon anyhow.

joestringer replied:
Hmm. Multiple paths for initialization is a bit awkward for some of the other reasons I've just posted. I think that this ends up making the code more complicated in a sort of "defensive coding" kind of way. If we can just guarantee that the structure is initalized early enough, then we don't need to worry about any of this extra logic.

christarazi replied:
We can guarantee that the DNS proxy in a non-test context will always initialize the LRU before serving DNS responses because we've put the LRU init inside the DNS proxy init.

What we can't really guarantee is if the daemon init ordering is for some reason shuffled around such that the K8s resources start syncing before the DNS proxy is up (and LRU is init'd). In that case, the LRU will not be initialized at first so we could be compiling regexes without caching them for a period of time, until the DNS proxy init starts. 

I'm not sure there is any way around this, at least until we have refectored the daemon init such that we can explicitly control which steps must happen before others, and so on.

This is why the original code was checking if the LRU was nil and just passing straight down to the regex compile call without doing anything else, because we assume that eventually, as long as we aren't in the testing context, that the LRU will get init'd.

Hopefully that clarifies this issue. So if the defensive coding is not really worth it, happy to remove it and go back to the original code as mentioned above.

joestringer replied:
> What we can't really guarantee is if the daemon init ordering is for some reason shuffled around such that the K8s resources start syncing before the DNS proxy is up (and LRU is init'd). In that case, the LRU will not be initialized at first so we could be compiling regexes without caching them for a period of time, until the DNS proxy init starts.

I was fishing for is a solution to this. If we don't provide this guarantee, then we could be spending more time/effort/memory on these regex compilations. For that undetermined period on startup, the cache does not work at all. Worst case, it makes the cache completely ineffective on startup and then it's only useful over time when new policies are added. My hope with the "error" suggestion would be that if the initialization returns an error up the stack, then at some point this will either trigger an agent-wide `log.Fatal()` or some sort of prominent error message to be logged. Then we catch this as part of testing in CI, detect whenever the guarantee is no longer provided, and make sure we fix it before that code gets into a release. That said, I'm not attached to this solution.

In my last comment, I was thinking that ideally there's just one place where we guarantee that the LRU is initialized. If we're going to make this Compile function more complicated to handle this case, then is it still useful to explicitly initialize the LRU from the DNS proxy? Presumably the first time that the DNS proxy uses this code, it will also call into the same Compile function. Is the DNS proxy initialization redundant, then? Maybe we can just contain the "worry about LRU initalization" logic to this Compile function?

(The other alternative is just explicitly initialize it super early even directly from the daemon init routines or something, but I'm not overly sold on that approach either).

christarazi replied:
Ok fair enough. Apologies this took longer than it should have. Yes returning an error will just simplify things. 

> My hope with the "error" suggestion would be that if the initialization returns an error up the stack, then at some point this will either trigger an agent-wide log.Fatal() or some sort of prominent error message to be logged. Then we catch this as part of testing in CI, detect whenever the guarantee is no longer provided, and make sure we fix it before that code gets into a release. That said, I'm not attached to this solution.

This didn't sound great to me at first because it's kind of a "stop the world" error, but it's not a bad idea now that I think about it a bit more. We won't run into this often anyway, if it all.

christarazi replied:
Implemented.

joestringer replied:
:+1: No need to apologise, it's good practice to have fully discussion like this to try to figure out the ideal implementation or challenge our assumptions. You also can't read my mind unless I put it into words first :)

joestringer commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,55 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile regex: %w", err)
+	}
+	RegexCompileLRU.Add(p, n)
```
I'm not seeing any form of locking here, how do we ensure this is safe between goroutines/threads? Does the underlying library provide any guarantees in this regard?

christarazi replied:
:scream: I somehow missed this. Good catch, I don't know why I didn't check my assumption on that. I assumed that a library like this would include its own locking.

```go
// Cache is an LRU cache. It is not safe for concurrent access.
type Cache struct {
```

joestringer replied:
Also worth noting in particular, that even if we protect the read/write of the LRU, the underlying entries are also being returned. So there's also the question of whether the underlying regex object can be concurrently used by multiple goroutines. (Though I'm guessing maybe we already thought about this in the existing design...?)

joestringer replied:
To answer the last question:

https://pkg.go.dev/regexp#Regexp

> A Regexp is safe for concurrent use by multiple goroutines, except for configuration methods, such as Longest.

joestringer commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -821,17 +819,12 @@ func GetSelectorRegexMap(l7 policy.L7DataMap, lru *lru.Cache) (CachedSelectorREE
+		case rei != nil && err == nil:
+			newRE[selector] = rei
+		case err != nil:
+			return nil, err
```
Any particular reason this doesn't follow the typical go form?

```suggestion
		rei, err := re.CompileRegex(mp)
		if err {
			return nil, err
		}
		newRE[selector] = rei
```

christarazi replied:
No, probably a leftover from iterating on the design of `CompileRegex`. Will simplify.

joestringer reviewed (commented):
I'm mainly unsure about the locking requirements particularly if this is to be used again from another goroutine now.

schlosna commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,92 @@
+type regexCompileLRU struct {
+	// The lru package doesn't provide any concurrency guarantees so we must
+	// provide our own locking.
+	*lock.Mutex
```
Relatedly, should this use a read/write mutex so that happy path cache hits only use read lock and don't block other readers, while cache miss acquires write lock?

```suggestion
	*lock.RWMutex
```

joestringer replied:
Same answer as here: https://github.com/cilium/cilium/pull/19632#discussion_r862384915

schlosna commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,92 @@
+	}
+	RegexCompileLRU.Lock()
+	r, ok := RegexCompileLRU.Get(p)
+	RegexCompileLRU.Unlock()
```
Should this take out just a read lock to avoid blocking on read only cache hit happy path?

```suggestion
	RegexCompileLRU.RLock()
	r, ok := RegexCompileLRU.Get(p)
	RegexCompileLRU.RUnlock()
```

joestringer replied:
I think that the concern here is that since the underlying structure is an LRU, presumably the `Get()` operation also modifies the underlying order that is stored, so readlocks would not be safe for concurrent access.

Overall though I presume that this will be primarily used during policy calculation and then the caller stores the underlying regex object for regular usage against the target requests without holding this lock, so it shouldn't be too bad even with a regular mutex.

christarazi replied:
^ Correct. If you look at the `Get()` call, you can see it's potentially modifying the underlying structure to evict the oldest member, etc.

schlosna replied:
Ahh, good call

joestringer commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,92 @@
+// i.e. compiling the FQDN rules or handling FQDN selectors). If that doesn't
+// hold true, then we attempt a best-effort to initialize the LRU.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	if RegexCompileLRU == nil {
```
I think that even the global check here should be protected by the mutex, to ensure that this pointer isn't modified by one goroutine while another goroutine is initializing it?

christarazi replied:
Good point, I've opted to use `atomic.{Load,Store}Pointer()`, instead of creating a mutex that will only be used once.

christarazi committed:
defaults: Define FQDN regex compile LRU size default

This value will be used during initialization of the LRU in case we
don't have the value from the user.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon, fqdn: Provide method for non-cached validation

In the previous commits, we've converged all paths to use the FQDN regex
compilation LRU. However, this can be problematic if we expose the LRU
via the API (over the CLI) because users can pass a pattern, i.e.
`cilium fqdn list -p <pattern>` which could needlessly evict useful
entries from the LRU.

extractDNSLookups() and deleteDNSLookups() are the only code paths that
are called from the API with a user-passed pattern that will be
compiled, therefore we define a new validation method to bypass the
LRU.

All other code paths will use matchPattern.Validate() which consults the
LRU.

Suggested-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// InitRegexCompileLRU initializes the singleton LRU cache.
+func InitRegexCompileLRU(size int) (*RegexCompileLRU, error) {
```
nit, but typically when Cilium code creates an object it will begin with `New...()`. We can touch this up before merge though.

```suggestion
// NewRegexCompileLRU initializes the singleton LRU cache.
func NewRegexCompileLRU(size int) (*RegexCompileLRU, error) {
```

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -933,7 +933,7 @@ func initializeFlags() {
 	option.BindEnv(option.FQDNProxyResponseMaxDelay)
 
-	flags.Int(option.FQDNRegexCompileLRUSize, 1024, "Size of the FQDN regex compilation LRU. Useful for heavy but repeated FQDN MatchName or MatchPattern use")
+	flags.Int(option.FQDNRegexCompileLRUSize, defaults.FQDNRegexCompileLRUSize, "Size of the FQDN regex compilation LRU. Useful for heavy but repeated FQDN MatchName or MatchPattern use")
```
Hmm wait, this description says that it configures the FQDN MatchName / MatchPattern regex LRU size, but it actually configures the DNS L7 rule MatchName / MatchPattern. The FQDN one is being hardcoded to the `defaults.FQDNRegexCompileLRUSize`.

twpayne commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,82 @@
+// LRU. This function will return an error if the LRU has not already been
+// initialized.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	lru := (*RegexCompileLRU)(atomic.LoadPointer(&regexCompileLRU))
```
Why is `atomic.LoadPointer` needed here?

christarazi replied:
See https://github.com/cilium/cilium/pull/19632#discussion_r862194361.

twpayne commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,82 @@
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	lru := (*RegexCompileLRU)(atomic.LoadPointer(&regexCompileLRU))
+	if lru == nil {
+		return nil, errors.New("FQDN regex compilation LRU not yet initialized")
```
Shouldn't this be a `panic` to catch misuse of this function early and loudly?

christarazi replied:
Using panic is not something that we have a clear pattern for doing in Cilium. Yes we do have lots of Panic lines, but I'd venture to say those were all likely added to help with unit testing.

In this case, I avoided panic because this could result in Cilium crashing in a user environment and grinds the usage of Cilium to a halt. In particular, this code is called triggered from the following locations:
* on policy add
* on endpoint regeneration (policy recalculation)
* from the built-in CLI (`cilium policy ...`)
  * which is similar to the first point, because they're both API

Rather, I opted for an error that the user can share with us instead and we can fix. It's basically just a less intrusive panic. 

I agree in principle that panic gives us faster feedback by being more drastic, but I'm not sure that it's appropriate in this case.

joestringer replied:
In general we have shied away from `panic()` throughout the codebase specifically for this kind of resiliency reason. It's easy to introduce these statements with the best intentions, but then for them to slip through testing and end up causing production incidents in live environments not because of the condition that occurred, but because the agent is panicking.

The main exception is where the user has explicitly configured conflicting options, there we panic early to give the user feedback that this combination of settings is unsupported. The idea there is that as soon as the user installs / enables the option, they should get immediate feedback about the conflict.

twpayne commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,82 @@
+	if lru == nil {
+		return nil, errors.New("FQDN regex compilation LRU not yet initialized")
+	}
+	lru.Lock()
```
Why not use a `sync.RWMutex` here? I saw another person made the same suggestion, but the suggestion was marked as resolved without comment or code change.

christarazi replied:
See https://github.com/cilium/cilium/pull/19632#discussion_r869385577.

(I think there was some confusion with the PR being opened for v1.9 backport immediately and people reviewing this vs that PR.)

twpayne commented on pkg/fqdn/re/re.go:
```diff
@@ -0,0 +1,82 @@
+func InitRegexCompileLRU(size int) error {
+	if size < 0 {
+		return fmt.Errorf("failed to initialize FQDN regex compilation LRU due to invalid size %d", size)
+	} else if size == 0 {
```
Should `size == 0` not disable the cache completely?

christarazi replied:
The default is non-zero, so it would take someone explicitly to configure this to 0 or a developer explicitly calling this function with 0. I was mostly just mirroring the behavior of the library, but adding a warning in case someone did this without thinking. Do you suggest this becomes an error instead?
